File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int [] res = new int[nums.length]; 

        for(int i = 0; i < res.length; i++){
            res[i] = 1;
        }

        int product = 1;
        for(int i = 1; i < res.length; i++){
            product *= nums[i - 1];
            res[i] = product * res[i];
        }

        product = 1;

        for(int i = res.length - 2; i >= 0; i--){
            product *= nums[i + 1];
            res[i] = product * res[i]; 
        }

        return res;
        
    }
}  

File: Data Structures & Algorithms/n-queens-ii/submission-0.js
class Solution {
    /**
     * @param {number} n
     * @return {number}
     */
    totalNQueens(n) {
        const cols = new Set();
        const pD = new Set(); // (r + c)
        const nD = new Set(); // (r - c)
        let res = 0;

        const backtrack = (r) => {
            if(r === n) {
                res++;
                return;
            }

            for(let c = 0; c < n; c++){
                if(cols.has(c) || pD.has(r + c) || nD.has(r - c)) continue;

                cols.add(c);
                pD.add(r + c);
                nD.add(r - c);

                backtrack(r + 1);

                cols.delete(c);
                pD.delete(r + c);
                nD.delete(r - c);

            }
        }

        backtrack(0);

        return res;
    }
}
